### ISSUE TITLE
Support For Blackmagic Camera Tally Through Arduino And Blackmagic 3G-SDI Arduino Shield

### ISSUE THREAD

**Issue by zackdickinson16** [josephdadams/TallyArbiter#233]
### What feature would you like to see added to TallyArbiter?

I'm looking for a way to use TallyArbiter with Arduino And Blackmagic 3G-SDI Arduino Shield to trigger the tally lights on black magic cameras such as the (Blackmagic Studio Camera 4K Pro), I need this because I'm using TriCaster and I'd like the built in tally lights to work on the cameras with TriCaster.

I'm guessing the Arduino would just have to act as a listener and listen to the TallyArbiter server running on the network.

I have all the pieces of software and libraries that I would like to use I just don't know how to program them all together to make them work with TallyArbiter.

Libraries and programs:
Arduino Ethernet Shield - **Ethernet.h & EthernetUdp ??**
Blackmagic 3G-SDI Arduino Shield - **BMDSDIControl.h**
TallyArbiter - pieces of the **TallyArbiter-M5StickCListener** ??

I just don't know how to put all these together or if there's an easier way?

Thanks,
Zack

**Comment by JTF4** [comment]
@josephdadams and I were looking at building this listener. Our biggest problem is that we don't have the hardware to develop with. 

One thing I would say is that I would use the raspberry pi hat before the arduino one. The raspberry pi will give you native Ethernet. 

Ill have a look at everything tomorrow and see what I can do though.

**Comment by zackdickinson16** [comment]
Let me know if I can help test anything with the hardware, Black magic Arduino code is pretty simple just high or low for light on or off and true or false four red or green. Of course that's probably not the hard part, 

I was having a hard time figuring out how to integrate TallyArbiter I'm guessing it with WebSocketsClient.h,
,SocketIOclient.h and Arduino_JSON.h but I don't know whole lot about "coding". I was going down a rat hole of trying to use OSC messages but didn't think that would be the best idea.

Unfortunately I don't think Black Magic supports raspberry pi or has any hardware for it?

Anything you guys can do is greatly appreciated!!

Thanks

**Comment by JTF4** [comment]
You're right! I thought they had one. 

Do you have a picture of your setup that you could send me for reference of your setup?

**Comment by zackdickinson16** [comment]
![IMG_2153](https://user-images.githubusercontent.com/57646814/137610741-b2068ecb-005a-428b-8164-85729855cbc9.jpg)
![IMG_2155](https://user-images.githubusercontent.com/57646814/137610744-3d81ce4a-e27d-47e5-bb4a-e9e0d71dda55.jpg)
![IMG_2160](https://user-images.githubusercontent.com/57646814/137610746-3ff3cbeb-8e64-4dc4-be69-9a6461264137.jpg

https://user-images.githubusercontent.com/57646814/137610814-82a94ce6-a150-491f-a8a4-a6c1dc27c3ca.mov

Arduino Uno Rev3
Blackmagic Studio Camera 4K Pro
Blackmagic 3G-SDI Arduino Shield
Ethernet Shield W5100

Let me know if you need anything else!!

**Comment by josephdadams** [comment]
I actually prototyped the code for this over a year ago. Never had the hardware to test on. I may not have kept it but I’ll dig around and see if I can find it. It will need refreshing anyway in light of our 3.0 architecture. Personally rather than implement a listener client per se I would create a program on the Arduino that subscribed to all tally information, similar to how we implemented VMix clients. @JTF4 we could do this over simple TCP rather than websockets.

@zackdickinson16 a remote computer we can access that is connected to the Arduino and shield, along with a webcam pointed at the monitor with the tally light would be extremely helpful in testing.

**Comment by MatteoGheza** [comment]
@JTF4 if you want to use REST api/UDP/etc let me known and I'll implement this in backend.
BTW, can you take a look at #225 since I don't have M5Stackc or M5Atom to test the updated firmware?

**Comment by JTF4** [comment]
@MatteoGheza yep! Do you think we could build a rest based api that allows for device reassigns from the main server? If we could do that, then I think we would have a great protocol for devices that don't play as well with socket.io

**Comment by MatteoGheza** [comment]
> @MatteoGheza yep! Do you think we could build a rest based api that allows for device reassigns from the main server? If we could do that, then I think we would have a great protocol for devices that don't play as well with socket.io 

I'm working on a new REST API.

**Comment by hrueger** [comment]
How would we push new tally data to a listener client with REST? Does the Device do polling or are we using something like [SSE](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events)?

**Comment by JTF4** [comment]
If I had to guess it's in error in one of the included libraries and not your direct code. I would start by looking at the actual code contents of the Black Magic library and seeing if it includes that.

**Comment by zackdickinson16** [comment]
I'm pretty sure it's the SocketIoClient.h Library?, but I'm not very good at this stuff so I wouldn't have a clue.
